--- a/include/boost/asio/detail/consuming_buffers.hpp
+++ b/include/boost/asio/detail/consuming_buffers.hpp
@@ -102,7 +102,7 @@

     std::advance(next, next_elem_);
     std::size_t elem_offset = next_elem_offset_;
-    while (next != end && max_size > 0 && result.count < result.max_buffers)
+    while (next != end && max_size > 0 && (result.count) < result.max_buffers)
     {
       Buffer next_buf = Buffer(*next) + elem_offset;
       result.elems[result.count] = boost::asio::buffer(next_buf, max_size);





